src/io/logwatcher.rs

```diff
@@ -1,121 +1,127 @@
+            }
+            let mut buff: Vec<u8> = Vec::new();
+            let _ = file.read_to_end(&mut buff);
 
-                // Check if file has been shortened
-                if self.reader.get_ref().metadata().unwrap().len() < self.pos {
-                    log::warn!("Console.log file was reset");
-                    self.pos = self.reader.get_ref().metadata().unwrap().len();
-                    self.last_activity = Instant::now();
-                }
+            let mut reader = Cursor::new(buff);
+            reader.set_position(self.cpos);
 
-                // Reopen the log file if nothing has happened for long enough in case the file has been replaced.
-                let time = Instant::now().duration_since(self.last_activity);
-                if time.as_secs() > 10 {
-                    log::debug!("No activity in 10 seconds, reopening.");
-                    let f = match File::open(&self.filepath) {
-                        Ok(x) => x,
-                        Err(_) => return None,
-                    };
+            if let Ok(mod_time) = file.metadata().unwrap().modified() {
+                Some(ReadMD(Some(mod_time), reader))
+            } else {
+                Some(ReadMD(None, reader)) // TODO: fix this grossness
+            }
+        } else {
+            None
+        }
+    }
 
-                    let metadata = match f.metadata() {
-                        Ok(x) => x,
-                        Err(_) => return None,
-                    };
+    fn read_lines(&mut self) {
+        if let Some(mut read_md) = self.seek_to_pos() {
+            self.last_read = read_md.0;
 
-                    let mut reader = BufReader::new(f);
-                    let pos = metadata.len();
-                    reader.seek(SeekFrom::Start(pos)).unwrap();
+            let mut data: Vec<u8> = Vec::new();
+            if let Ok(bytes_read) = read_md.1.read_to_end(&mut data) {
+                self.cpos += bytes_read as u64;
+            } else {
+                log::error!("Could not read file data!");
+                return;
+            }
 
-                    self.pos = pos;
-                    self.reader = reader;
-                    self.last_activity = Instant::now();
-                    return None;
+            let mut lines: VecDeque<String> = VecDeque::new();
+            // Gross predicate to split on the u8 repr of a newline.
+            let line_iter = data.split(|&x| x == 0xAu8);
+            for byte_str in line_iter {
+                let s = String::from_utf8_lossy(byte_str);
+                if s.is_empty() || s.trim().is_empty() {
+                    continue;
                 }
-
-                self.reader.seek(SeekFrom::Start(self.pos)).unwrap();
-                return None;
-            }
-            Err(err) => {
-                log::error!("Logwatcher error: {}", err);
+                lines.push_back(s.to_string());
             }
```

Comment:
Instead of splitting the lines yourself, if you convert the whole thing into a String you can use the standard library String's `lines` method to split it, this will also handle the case of the lines being terminated by different endings. (`\n` vs `\r\n`).

I would change it to something like:
```rust
let s = String::from_utf8_lossy(data);
self.lines_buf = s.lines().cloned().collect();
```
(or whatever is similar that would compile.)

Reply:
The best outcome would be to entirely leave it as a byte stream and just pass byte streams around then only cast to string when we need it (i.e. on location), but that would require reworking other portions of code. What do you think of that?

I should note that even with stripping failed utf8, it still seems to maintain a lot of the supported non-utf8 characters that steam puts through. I.e. I still catches unicode emojis and all sorts of fun things. I'm not sure exactly how this is working?

Reply:
> The best outcome would be to entirely leave it as a byte stream and just pass byte streams around then only cast to string when we need it (i.e. on location), but that would require reworking other portions of code. What do you think of that?

I don't think that would be the best outcome. If we can't just drop in `OsString` as a replacement then we would want to be using some UTF16 supported String type like from an external crate. 

I would have to double check and give it a go but I'm still not convinced `OsString` can't easily do that, as the documentation talks about it being in some intermediate representation that isn't strictly UTF16 to share more stuff between OS implementations, so I imagine all the functionality needed would be available across all OSes. Not actually sure though, never really used them before.

Reply:
But I think that's a worry for later, I'm happy with how this is currently, and it is closer to what I imagined initially.

Reply:
Dealing with OsStrings has some jank attached if im not mistaken:
```rs
#[cfg(target_os = "linux")]
mod UTF16Support {
    use std::os::unix::ffi::OsStringExt;
    pub fn from_bytes(buf: &[u8]) -> OsStringExt {
        OsString::from_vec(&buf.to_vec())
    }
}

#[cfg(target_os = "windows")]
 mod UTF16Support {
    use std::{os::windows::ffi::OsStringExt, ffi::OsString};
    pub fn from_bytes(buf: &[u8]) -> OsString {
        let stream: Vec<u16> = buf
            .chunks_exact(2)
            .into_iter()
            .map(|a| u16::from_ne_bytes([a[0], a[1]]))
            .collect();
        OsString::from_wide(&stream)
    }
}
```

Reply:
Also that windows function assumes a certain endianness of the data, and currently doesn't check alignment

Reply:
Ah I see, so the Windows and Linux ones expect the data in a different format in the first place. Yeah I guess that does make sense. We would probably want to use an external crate then for that support, which I don't think is worth it at the moment if we can safely coerce the data into a compliant format without losing the important stuff.

Reply:
that func didnt work anyway, new commit does it different but still has variable modules

Reply:
I'm not sure this is preferable over the original one that stripped any unsupported characters. Another issue is that this is now explicitly for Windows and Linux, and will require code changes if someone wanted to compile for MacOS (doubtful, but still possible, and I would rather maintain compatibility if possible. I ended up making my own solution when Pazer BD didn't support Linux after all). Once I test it I would be happy to merge the first few commits from this branch but I'm hesitant about these new commits.

Reply:
If it ends up being beneficial going the whole way to support UTF16 I would rather delegate any platform-specific stuff to an external crate, particularly if it will require any unsafe code. Provided there is a good one out there.

Tbh though I don't think we will need full UTF16 support, and UTF8 with invalid characters stripped will probably suffice, at least for now.

Reply:
well so the only unsafe happening here is an unsafe cast to a string that doesn't check the bytes are actually valid utf8, which is actually what we want when passing that string into the 'from_str' function. We can roll back to original, we can just throw the bytes into a `String::from_utf8_lossy()` and  deal with the loss

Reply:
for macOS support it behaves the same os linux, so the target_os could be shared there

Reply:
Yeah okay, if you are quite confident that this way works and you want to do it this way then make sure to test it thoroughly. Also make sure to change the platform-specific config to not-windows instead so it includes Mac (an example of this is shown in the settings or wherever the TF2 directory is chosen, if you are unsure how).

I'm slightly more inclined to stick with the old lossy one but I'm willing to accept this one if you are sure it works and make it more compatible.